src/probnum/linalg/solvers/policies/_conjugate_gradient.py

```diff
@@ -11,21 +14,101 @@ class ConjugateGradientPolicy(_linear_solver_policy.LinearSolverPolicy):
     r"""Policy returning :math:`A`-conjugate actions.
 
     Selects the negative gradient / residual as an initial action :math:`s_0 = b - A x_0` and then successively generates :math:`A`-conjugate actions, i.e. the actions satisfy :math:`s_i^\top A s_j = 0` iff :math:`i \neq j`.
+
+    Parameters
+    ----------
+    reorthogonalization_fn_residual
+        Reorthogonalization function, which takes a vector, an orthogonal basis and optionally an inner product and returns a reorthogonalized vector. If not `None`
+        the residuals are reorthogonalized before the action is computed.
+    reorthogonalization_fn_action
+        Reorthogonalization function, which takes a vector, an orthogonal basis and optionally an inner product and returns a reorthogonalized vector. If not `None`
+        the computed action is reorthogonalized.
     """
 
+    def __init__(
+        self,
+        reorthogonalization_fn_residual: Optional[
+            Callable[
+                [np.ndarray, Iterable[np.ndarray], linops.LinearOperator], np.ndarray
+            ]
+        ] = None,
+        reorthogonalization_fn_action: Optional[
+            Callable[
+                [np.ndarray, Iterable[np.ndarray], linops.LinearOperator], np.ndarray
+            ]
+        ] = None,
+    ) -> None:
+        self._reorthogonalization_fn_residual = reorthogonalization_fn_residual
+        self._reorthogonalization_fn_action = reorthogonalization_fn_action
+
     def __call__(
         self, solver_state: "probnum.linalg.solvers.LinearSolverState"
     ) -> np.ndarray:
 
         action = -solver_state.residual.copy()
 
+        if self._reorthogonalization_fn_residual is not None and solver_state.step == 0:
+            solver_state.cache["reorthogonalized_residuals"] = [solver_state.residual]
+
         if solver_state.step > 0:
-            # A-conjugacy correction (in exact arithmetic)
-            beta = (
-                np.linalg.norm(solver_state.residual)
-                / np.linalg.norm(solver_state.residuals[solver_state.step - 1])
-            ) ** 2
+            # Reorthogonalization of the residual
+            if self._reorthogonalization_fn_residual is not None:
+                residual, prev_residual = self._reorthogonalized_residuals(
+                    solver_state=solver_state
+                )
+            else:
+                residual = solver_state.residual
+                prev_residual = solver_state.residuals[solver_state.step - 1]
 
+            # A-conjugacy correction (in exact arithmetic)
+            beta = (np.linalg.norm(residual) / np.linalg.norm(prev_residual)) ** 2
             action += beta * solver_state.actions[solver_state.step - 1]
 
```

Comment:
Is line 65 correct? In CG, the next search direction is computed as s_{i+1} = r_{i+1} + beta s_{i}. Here it appears to be computed as s_{i+1} = s_i + bets s_i. This line is not changed in this pull request, so it is probably not an error.

Reply:
Should be, since action is set to `-solver_state.residual.copy()` at the beginning of the method. Or do we need the reorthogonalized residual here?

Reply:
It is correct on the first step because the first action is the negative residual. In CG, the action would typically be updated as 
```
beta = (np.linalg.norm(residual) / np.linalg.norm(prev_residual)) ** 2
action = -residual + beta * solver_state.actions[solver_state.step - 1]
```

Even though it is different here, it may not cause problems because the belief mean is updated as
```
cov_xy = solver_state.belief.x.cov @ action_A.T
gram = action_A @ cov_xy + self._noise_var
gram_pinv = 1.0 / gram if gram > 0.0 else 0.0
gain = cov_xy * gram_pinv
mean=solver_state.belief.x.mean + gain * proj_resid,
``` 
instead of
```
cov_xy = ORIGINAL_COVARIANCE @ action_A.T
gram = action_A @ cov_xy + self._noise_var
gram_pinv = 1.0 / gram if gram > 0.0 else 0.0
gain = cov_xy * gram_pinv
mean=solver_state.belief.x.mean + gain * proj_resid,
``` 
Using the most recent covariance acts like a reorthogonalization, so it may make both versions of the action update act the same. I am not sure about this so I will think about it more.

Reply:
Good catch @treid5 . It does compute the update step as `-residual + beta * beta * solver_state.actions[solver_state.step - 1]` since in line 48 `action` is set to be the current residual (as to mirror gradient descent). However, now that i've added the reorthogonalization this residual currently is actually *not* reorthogonalized, while the ones used for the conjugacy correction are. I'll fix that.

Reply:
Good catch @treid5 . It does compute the update step as -residual + beta * beta * solver_state.actions[solver_state.step - 1] since in line 48 action is set to be the current residual in every iteration (as to mirror gradient descent). However, now that i've added the reorthogonalization this residual currently is actually not reorthogonalized, while the ones used for the conjugacy correction are. I'll fix that.